src/utils/ucc_string.c

```diff
@@ -0,0 +1,102 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2021.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+#include "ucc_string.h"
+#include "ucc_malloc.h"
+#include "ucc_log.h"
+#include <ctype.h>
+#include <ucs/sys/string.h>
+char **ucc_str_split(const char *str, const char *delim)
+{
+    unsigned alloc_size = 8;
+    unsigned size       = 0;
+    char   **out;
+    char    *str_copy, *token, *saveptr;
+    int      i;
+    out = ucc_malloc(alloc_size * sizeof(char *), "str_split");
+    if (!out) {
+        ucc_error("failed to allocate %zd bytes for str_split",
+                  alloc_size * sizeof(char *));
+        return NULL;
+    }
+    str_copy = strdup(str);
+    if (!str_copy) {
+        ucc_error("failed to duplicate string");
+        goto error;
+    }
+    token = strtok_r(str_copy, delim, &saveptr);
+    while (NULL != token) {
+        out[size] = strdup(token);
+        if (!out[size]) {
+            ucc_error("failed to duplicate string");
+            goto error;
+        }
+        size++;
+        if (size == (alloc_size - 1)) { /* keep 1 for NULL mark */
+            alloc_size *= 2;
+            out = ucc_realloc(out, alloc_size * sizeof(char *), "str_split");
+            if (!out) {
+                ucc_error("failed to reallocate %zd bytes for str_split",
+                          alloc_size * sizeof(char *));
+                goto error;
+            }
+        }
+        token = strtok_r(NULL, delim, &saveptr);
+    }
+    out[size] = NULL;
+    ucc_free(str_copy);
+    return out;
+error:
+    for (i = 0; i < size; i++) {
+        ucc_free(out[i]);
+    }
+    ucc_free(out);
+    ucc_free(str_copy);
+    return NULL;
+}
+
+unsigned ucc_str_split_count(char **split)
```

Comment:
ucc_str_split_count -> ucc_str_count ?

Reply:
it is technically not a string (str) but a string split - ie. array of strings terminated with a NULL entry. So, count and free below refer to that and hence called "split_count/free".